### PR TITLE
[ntuple] Add tutorial for multi-threaded filling

### PR DESCRIPTION
**Pull request by jblomer**
# This Pull request:
Add `RNTupleWriter::CreateEntry()` and uses it in a demonstration on multi-threaded writing.

### CODE REVIEW

**Comment by jblomer** [comment]
@pikacic This could be one way for multi-threaded writing. I think we have still some issues regarding the life time of the memory regions used for filling.  In this example, `REntry::Get` returns a raw pointer (not a shared pointer like `RNTupleModel::MakeField`).  But the basics seems to work.  Please let me know if you have comments!

File: tree/ntuple/v7/inc/ROOT/RNTuple.hxx [hunk 1]
```diff
@@ -392,6 +392,8 @@ public:
    /// Ensure that the data from the so far seen Fill calls has been written to storage
    void CommitCluster();
 
+   std::unique_ptr<REntry> CreateEntry() { return fModel->CreateEntry(); }
```
Comment (pcanal) [inline tree/ntuple/v7/inc/ROOT/RNTuple.hxx:395]: This augment the interface and probably deserve its own commit, doesn't it?

File: tutorials/v7/ntuple/ntpl007_mtFill.C [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+// Number of events to generate is kNEventsPerThread * kNWriterThreads
+constexpr int kNEventsPerThread = 25000;
+
+// Thread function to generate and write events
+void FillData(std::unique_ptr<REntry> entry, RNTupleWriter *ntuple) {
+   // Protect the ntuple->Fill() call
+   static std::mutex gLock;
+
+   static std::atomic<std::uint32_t> gThreadId;
+   auto threadId = gThreadId++;
```
Comment (pcanal) [inline tutorials/v7/ntuple/ntpl007_mtFill.C:57]:
```suggestion
   const auto threadId = gThreadId++;
```

File: tutorials/v7/ntuple/ntpl007_mtFill.C [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+         float px, py, pz;
+         prng->Rannor(px, py);
+         pz = px*px + py*py;
+
+         vpx->emplace_back(px);
+         vpy->emplace_back(py);
+         vpz->emplace_back(pz);
+      }
+
+      std::lock_guard<std::mutex> guard(gLock);
```
Comment (pcanal) [inline tutorials/v7/ntuple/ntpl007_mtFill.C:85]: As a side questions, is this (user locking call to Fill) the final interface or will the lock/use-of-atomics get added to RNTuple later?

**Review by pcanal — APPROVED** [review]

**Comment by pikacic** [comment]
I had a look ant it seems very nice.

About the discrepancy between `REntry::Get` and `RNTupleModel::MakeField`, one may argue that `MakeField` might return a raw pointer too as the memory of the field is owned by the model. OTOH, while it's kind of easy to keep track of the lifetime of an `REntry` (I get it, I use it, I drop it), it seems more difficult to keep track of `RNTupleModel` lifetime, which is bound to the lifetime of `RNTupleWriter`.
Using raw pointers from `MakeField` is fine if we assume that the writer of the code only accesses the fields within the scope of `RNTupleWriter` (but the fields are accessible before the writer is created... more confusing).

My personal opinion is that the `CreateEntry` way is the best option for the single thread case too:
- I define a model
- create a writer based on the model
- get the *buffer* (`REntry`) to write to (one per thread, for example, even if I have only one thread)
- commit the *buffer* to the writer

In this way I own the `REntry`, but it has a layout in memory that is directly understood by the serialization process without the need of extra copies.

We can also think of a way of constructing the writer from an *inlined* model, without the need for repeated calls to `MakeField`. Something like:
```cpp
auto ntuple = RNTupleWriter::Recreate({
    Field<std::uint32_t>("id"),
    Field<std::vector<float>>("vpx"),
    Field<std::vector<float>>("vpy"),
    Field<std::vector<float>>("vpz")
}, "NTuple", kNTupleFileName);

auto entry = ntuple->CreateEntry();
// cache the pointer for faster access in single thread,
// but I could use TLS (or a framework equivalent) for multithreading
auto& vpx = *entry->Get<std::vector<float>>("vpx");
for(auto& event: all_events) {
  vpx.clear();
  vpx.push_back(42.);
  ntuple->Fill(entry); // I would prefer ntuple->Write(entry) but it's a matter of taste
}
```

File: tutorials/v7/ntuple/ntpl007_mtFill.C (hunk 3 above)
Comment (pikacic) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: I actually have the same question. I would prefer the lock is internal, but I imagine we would need a non synchronized version for the single thread case.

A was also thinking about a non blocking version of `Fill` (e.g. `RNTuple::Writer(std::unique_ptr<REntry>)` that would enqueue the serialization for a dedicated thread. Of course you have to pay for the allocation/deallocation of `REntry` at every event:
```cpp
for(auto& event: all_events) {
  auto entry = ntuple->CreateEntry();
  auto &vpx = *entry->Get<std::vector<float>>("vpx");
  vpx.clear();
  vpx.push_back(42.);
  ntuple->Fill(std::move(entry));
}
```

File: tree/ntuple/v7/inc/ROOT/RNTuple.hxx (hunk 1 above)
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RNTuple.hxx:395]: Done

File: tutorials/v7/ntuple/ntpl007_mtFill.C (hunk 3 above)
Comment (jblomer) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: I was planning to keep `Fill()` unlocked for consistency reasons. While we could protect `Fill()`, the rest of the `RNTupleWriter` would still not be prepared to be called concurrently. I think it is cleaner to keep the class [conditionally thread safe](https://root.cern/manual/thread_safety/), like most other RNTuple classes.

Moving the REntries into a serialization queue is a really interesting idea! I don't think we need/should to address this in the `RNTupleWriter` class though. What we essentially need is a bounded, thread-safe queue for the producer-consumer thread synchronization task. I wonder if an implementation of such a queue would make sense as a generally available utility class in ROOT, @pcanal?

That said, using a dedicated thread for the serialization would only make sense if we actually spend a non-negligible amount of time in copying data from the in-memory objects into the I/O buffers.  For compressing I/O buffers, we already have infrastructure in place to do it multi-threaded.  That helps substantially for simple events but it is, as far as I can see, not the main bottleneck for writing/creating complex events.

@pikacic Perhaps we can discuss it in person?

Comment (pcanal) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: "bounded, thread-safe queue for the producer-consumer thread synchronization task": yes, it does sound useful.  The challenging part is the "bounded" part and more particularly what to do when the bound is reached.

**Comment by jblomer** [comment]
I very much like the idea of being able to define a model inline! See #8711 

What the `auto fldXyz = MakeField<type>("xyz")` approach is trying to avoid is having to spell the schema more than once.  As schemas get large, it can be a little cumbersome to remember and write down all the different types and field names in different places of the program. I do see, however, that the approach complicates reasoning about the lifetimes of the different variables.

File: tutorials/v7/ntuple/ntpl007_mtFill.C (hunk 3 above)
Comment (jblomer) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: I have in mind an implementation with conditional variables, so enqueuing would block if the queue is full.

Comment (pcanal) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: In some context (CMSSW for example), they have argued that this is sub-optiomal as it prevents the calling thread from doing other work while waiting and in similar context (eg TFileBufferMerger, we have to provide feedback on how full the queue was, some of the discussion also talked about a call back of sort). [@Dr15Jones @dan131riley might have more details]

Comment (Dr15Jones) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: So it looks to me that in the example, the most time consuming call would actually be `tuple->Fill(*entry);`? If so, then you would get very little concurrency. If the `Fill` call could make use of available threads (much like `TTree::Fill` if ImplicitMT is on) then you'd still have a problem since you'd have lots of possible threads all being blocked and therefore unusable by the `Fill` call. That would mean you'd have to oversubcribed the system, which isn't really a good idea.

An alternative design used frequently in CMSSW is to have a `SerialTaskQueue` with an associated callback for when the task finishes. The callback is used to asynchronously start another series of tasks. This is often done recursively: an asynchronous function setup work, when the work completes it enqueues a finishing task and then when the finishing task runs its last step is to call the asynchronous function which had started the whole process.

Below is a rewrite of the `Write()` function using this idea (it depends on code similar to https://github.com/cms-sw/cmssw/blob/master/FWCore/Concurrency/interface/SerialTaskQueue.h) using tbb::task_group to handle the thread pool. In the code, no TBB threads are ever blocked so they are always available for work (e.g. by `Fill`).

```C++
void nextEntryAsync(tbb::task_group& group, std::atomic<int>& countDown, int threadID, REntry& entry, RNTupleWriter& tuple, SerialTaskQueue& queue, TRandom3&rand) {

	group.run([]() {
		if(--countDown >=0) {
      auto id = entry->Get<std::uint32_t>("id");
      auto vpx = entry->Get<std::vector<float>>("vpx");
      auto vpy = entry->Get<std::vector<float>>("vpy");
      auto vpz = entry->Get<std::vector<float>>("vpz");
			
      vpx->clear();
      vpy->clear();
      vpz->clear();
      *id = threadID;
      
      int npx = static_cast<int>(prng->Rndm(1) * 15);
      // Set the field data for the current event
      vpx->reserve(npx);
      vpy->reserve(npy);
      vpz->reserve(npz);
      for (int j = 0; j < npx; ++j) {
          float px, py, pz;
          prng->Rannor(px, py);
          pz = px*px + py*py;

          vpx->emplace_back(px);
          vpy->emplace_back(py);
          vpz->emplace_back(pz);
      }      
      
      //only one thread at a time can fill the tuple
      // once fill is done, try to process next entry
      queue.push(group, [&,threadID]() {
        //internally, Fill could use available TBB threads
        tuple.Fill(entry);
        nextEntryAsync(group, countDown, threadID, entry, tuple, queue, rand);
      });
		}
	});
}

void Write() {
  
  // Create the data model
  auto model = RNTupleModel::Create();
  model->MakeField<std::uint32_t>("id");
  model->MakeField<std::vector<float>>("vpx");
  model->MakeField<std::vector<float>>("vpy");
  model->MakeField<std::vector<float>>("vpz");

  // We hand-over the data model to a newly created ntuple of name "NTuple", stored in kNTupleFileName
  auto ntuple = RNTupleWriter::Recreate(std::move(model), "NTuple", kNTupleFileName);
  
  //set the number of threads total as well as the number of threads for the local arena
  // the global setting makes sure any other arenas used will have their max threads restricted
  tbb::global_control control(tbb::global_control::max_allowed_parallelism, nWriterThreads);
  tbb::task_arena arena(nWriterThreads);
  arena.execute([&]() {
  
    std::atomic<int> countDown = kNEvents;
    
    SerialTaskQueue queue;
    
    std::vector<std::unique_ptr<REntry>> entries;
    std::vector<std::unique_ptr<TRandom3>> rands;
    entries.reserve(nWriterThreads);
    rands.reserve(nWriterThreads);
    
    tbb::task_group group;
    for(int i=0; i< nWriterThreads; ++i) {
      entries.emplace_back(ntuple->CreateEntry());
      rands.emplace_back(std::make_unique<TRandom3>());
      rands.back()->SetSeed();
      nextEntryAsync(group, countDown, i, *entries.back(), ntuple, queue, *rands.back());
    }
    group.wait();
  });
}
```

Comment (Dr15Jones) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: One item of note in this design, the SerialTaskQueue is implicitly bound because at most nWriteThreads-1 tasks can be waiting in the queue at any one time. That is because new tasks can only be spawned after a task has finished in the queue.

Comment (Dr15Jones) [reply tutorials/v7/ntuple/ntpl007_mtFill.C:85]: So pushing things a bit further using the 'task chaining' code (seen here https://github.com/cms-sw/cmssw/blob/master/FWCore/Concurrency/test/test_catch2_WaitingTaskChain.cc) the part of `Write()` that falls within the `arena.execute(...)` could probably be written as

```C++
std::atomic<int> countDown = 0;

SerialTaskQueue queue;

//setup per 'lane' data structures
std::vector<std::unique_ptr<REntry>> entries;
std::vector<std::unique_ptr<TRandom3>> rands;
entries.reserve(nWriterThreads);
rands.reserve(nWriterThreads);

tbb::task_group group;
FinalWaitingTask finalTask;
{
  //this ensures the finalTask can't be run until after the loop finishes
  WaitingTaskHolder final(group, finalTask);

  //create as many 'lanes' as threads
  for(int i=0; i< nWriterThreads; ++i)) {
    entries.emplace_back(ntuple->CreateEntry());
    auto entry = entries.back().get();
    rands.emplace_back(std::make_unique<TRandom3>());
    auto rand = rands.back().get();
    rand->SetSeed();
    
    //here is the 'chain' of tasks to be exetued
    chain::while([&count, kNEvents](){return ++count< kNEvents;},
        chain::first([&](auto iTask){ fillEntry(*entry); } ) 
        | chain::then([&](auto iTask){ queue.push([iTask]() { tuple->Fill(*entry); })})
    )   
    | runLast(final);
  }
}

group.wait();
```

The `chain::while` ability doesn't exist in the present code, but I believe it would be possible to create.
